File: src/ParentProgenyImputer.cpp
#include <cmath>
#include "../include/ParentProgenyImputer.h"
#include "../include/common.h"

using namespace std;

ParentProgenyImputer::ParentProgenyImputer(const vector<VCFFamilyRecord *>& rs,
							 bool is_mat_imputed_,
							 const vector<vector<int>>& ref_hs,
							 const Map& map_, double w) :
					VCFHMM(rs, map_, w), ref_records(rs), ref_haps(ref_hs),
					prev_h_table(collect_possible_previous_hidden_states()),
					is_mat_imputed(is_mat_imputed_),
					Cc(calc_Cc(rs)), Cp(calc_Cp(rs)) { }

vector<double> ParentProgenyImputer::calc_Cc(
							const vector<VCFFamilyRecord *>& rs) const {
	const size_t	M = rs.size();
	vector<double>	Cc(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cc[i] = Map::Kosambi(dist(rs[i], rs[i+1]));
	}
	return Cc;
}

vector<double> ParentProgenyImputer::calc_Cp(
							const vector<VCFFamilyRecord *>& rs) const {
	const double	K = 5.0;
	const size_t	M = rs.size();
	vector<double>	Cp(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cp[i] = Map::Kosambi(dist(rs[i], rs[i+1]) * K);
	}
	return Cp;
}

double ParentProgenyImputer::emission_probability(std::size_t i, int h, int op,
												const std::vector<int>& ocs,
												int phased_parent_gt) const {
	const size_t	N = num_progenies();
	const int	hp = h >> (N*2);
	const int	hp1 = hp % NH();
	const int	hp2 = hp / NH();
	const int	non_phased_parent_gt = ref_haps[hp1][i] |
										ref_haps[hp2][i] << 1;
	const int	mat_gt = is_mat_imputed ? phased_parent_gt :
												non_phased_parent_gt;
	const int	pat_gt = is_mat_imputed ? non_phased_parent_gt :
												phased_parent_gt;
	const double	Ep = E[non_phased_parent_gt][op];
	double	Ec = 0.0;
	for(size_t j = 0; j < N; ++j) {
		const int	hc = (h >> (j * 2)) & 3;
		const int	hc1 = hc & 1;
		const int	hc2 = hc >> 1;
		const int	gtc = gt_by_haplotypes(hc1, hc2, mat_gt, pat_gt);
		Ec += E[gtc][ocs[j]];
	}
	return Ep + Ec;
}

double ParentProgenyImputer::progeny_transition_probability(size_t i,
													int prev_hc, int hc) const {
	const int	t = prev_hc ^ hc;
	const double	cc = Cc[i-1];
	double	Tc = 0.0;
	for(size_t j = 0; j < num_progenies() * 2; ++j) {
		const int	tc = (t >> j) & 1;
		Tc += log(tc != 0 ? cc : 1.0 - cc);
	}
	return Tc;
}

double ParentProgenyImputer::parent_transition_probability(size_t i,
													int prev_hp, int hp) const {
	const double	cp = Cp[i-1];
	const int	hp1 = hp % NH();
	const int	hp2 = hp / NH();
	const int	prev_hp1 = prev_hp % NH();
	const int	prev_hp2 = prev_hp / NH();
	return log(prev_hp1 != hp1 ? cp : 1.0 - cp) +
		   log(prev_hp2 != hp2 ? cp : 1.0 - cp);
}

double ParentProgenyImputer::transition_probability(size_t i,
													int prev_h, int h) const {
	const size_t	N = num_progenies();
	const int	prev_hc = prev_h & ((1 << (N*2)) - 1);
	const int	hc = h & ((1 << (N*2)) - 1);
	const int	prev_hp = prev_h >> (N*2);
	const int	hp = h >> (N*2);
	return progeny_transition_probability(i, prev_hc, hc) +
			parent_transition_probability(i, prev_hp, hp);
}

vector<ParentProgenyImputer::DP> ParentProgenyImputer::initialize_dp() const {
	const size_t	L = NH() * NH() << (2*num_progenies());
	vector<DP>	dp(M(), DP(L, pair<double, int>(MIN_PROB, 0)));
	const VCFFamilyRecord	*record = ref_records[0];
	const size_t	phased_index = is_mat_imputed ? 0 : 1;
	const size_t	non_phased_index = is_mat_imputed ? 1 : 0;
	const int	phased_parent_gt = record->get_geno()[phased_index] & 3;
	const int	op = record->unphased(non_phased_index);
	vector<int>	ocs;	// observed progenies' genotypes
	for(size_t ic = 0; ic != num_progenies(); ++ic) {
		ocs.push_back(record->unphased(ic+2));
	}
	for(int h = 0; h < (int)L; ++h) {	// hidden state
		const double	E_all = emission_probability(0, h, op, ocs,
														phased_parent_gt);
		dp[0][h] = make_pair(E_all, h);
	}
	return dp;
}

// Collect possible previous hidden states for the hidden state.
vector<vector<int>>
ParentProgenyImputer::collect_possible_previous_hidden_states() const {
	const size_t	N = num_progenies();
	const size_t	L = NH() * NH() << (2*N);
	const size_t	Lc = 1 << (2*N);	// number of children states
	vector<vector<int>>	prev_h_table(L, vector<int>());
	for(int h = 0; h < (int)L; ++h) {	// hidden state
		const int	hc = h & (Lc - 1);
		const int	hp = h >> (N*2);
		const int	hp1 = hp % NH();
		const int	hp2 = hp / NH();
		// Neither side will crossover
		vector<int>	prev_hps;
		for(int h1 = 0; h1 < (int)NH(); ++h1) {
			const int	prev_h1 = h1 + hp2 * NH();
			prev_hps.push_back(prev_h1);
		}
		for(int h2 = 0; h2 < (int)NH(); ++h2) {
			if(h2 == hp2)	// for duplication
				continue;
			const int	prev_h1 = hp1 + h2 * NH();
			prev_hps.push_back(prev_h1);
		}
		
		vector<int>	prev_hcs;
		for(int hc1 = 0; hc1 < 1 << (N*2); ++hc1) {
			const int	t = hc ^ hc1;	// each bit means whether crossover
			int	num_crossovers = 0;
			for(int k = 0; k < (int)N*2; ++k) {
				if(((t >> k) & 1) == 1)
					num_crossovers += 1;
			}
			if(num_crossovers <= 1)
				prev_hcs.push_back(hc1);
		}
		
		for(auto p = prev_hps.begin(); p != prev_hps.end(); ++p) {
			for(auto q = prev_hcs.begin(); q != prev_hcs.end(); ++q) {
				if(*p == hp || *q == hc) {
					const int	prev_h = (*p << (N*2)) | *q;
					prev_h_table[h].push_back(prev_h);
				}
			}
		}
	}
	
	return prev_h_table;
}

int ParentProgenyImputer::compute_non_phased_parent_gt(int h, size_t i) {
	const size_t	N = this->num_progenies();
	const int	hp = h >> (N*2);
	const int	hp1 = hp % this->NH();
	const int	hp2 = hp / this->NH();
	return this->ref_haps[hp1][i] | (this->ref_haps[hp2][i] << 1);
}

void ParentProgenyImputer::update_dp(size_t i, vector<DP>& dp) const {
	const size_t	N = num_progenies();
	const size_t	L = NH() * NH() << (2*N);
	const VCFFamilyRecord	*record = ref_records[i];
	const size_t	phased_index = is_mat_imputed ? 0 : 1;
	const size_t	non_phased_index = is_mat_imputed ? 1 : 0;
	const int	phased_parent_gt = record->get_geno()[phased_index] & 3;
	const int	op = record->unphased(non_phased_index);
	vector<int>	ocs;	// observed progenies' genotypes
	for(size_t ic = 0; ic != num_progenies(); ++ic) {
		ocs.push_back(record->unphased(ic+2));
	}
	
	for(int h = 0; h < (int)L; ++h) {
		const double	E_all = emission_probability(i, h, op, ocs,
														phased_parent_gt);
		
		const auto&	prev_hs = prev_h_table[h];
		for(auto p = prev_hs.begin(); p != prev_hs.end(); ++p) {
			const int	prev_h = *p;
			
			const double	T = transition_probability(i, prev_h, h);
			const double	prob = dp[i-1][prev_h].first + (T + E_all);
			dp[i][h] = max(dp[i][h], make_pair(prob, prev_h));
		}
	}
}

void ParentProgenyImputer::update_genotypes(const vector<int>& hs) {
	const size_t	N = num_progenies();
	for(size_t i = 0; i < this->M(); ++i) {
		VCFFamilyRecord	*record = records[i];
		int	mat_gt = 0;
		int	pat_gt = 0;
		if(this->is_mat_imputed) {
			pat_gt = this->compute_non_phased_parent_gt(hs[i], i);
			record->set_geno(1, pat_gt | 4);
			mat_gt = record->get_geno()[0] & 3;
		}
		else {
			mat_gt = this->compute_non_phased_parent_gt(hs[i], i);
			record->set_geno(0, mat_gt | 4);
			pat_gt = record->get_geno()[1] & 3;
		}
		for(size_t j = 0; j < N; ++j) {		// each progeny
			const int	hc = hs[i] >> (j * 2);
			const int	hc1 = hc & 1;
			const int	hc2 = (hc >> 1) & 1;
			const int	gtc_int = gt_by_haplotypes(hc1, hc2, mat_gt, pat_gt);
			record->set_geno(j+2, gtc_int | 4);
		}
	}
}

void ParentProgenyImputer::impute() {
	// DP
	vector<DP>	dp = initialize_dp();
	for(int i = 1; i < (int)this->M(); ++i) {
		update_dp(i, dp);
	}
	
	const vector<int>	hs = trace_back(dp);
	update_genotypes(hs);
}

File: python/ParentProgenyImputer.py
from __future__ import annotations

# coding: utf-8
# ParentProgenyImputer.py
# 親がphasedのときに後代をimputeする

from math import log
from typing import List, Tuple

from VCFHMM import *
from Genotype import Genotype


#################### ParentProgenyImputer ####################

MIN_PROB = -1e300

class ParentProgenyImputer(VCFHMM[VCFFamilyRecord]):
	DP = List[Tuple[float, int]]	# (log of probability, prev h)
	
	def __init__(self, records: list[VCFFamilyRecord],
							ref_haps: list[list[int]],
							is_mat_imputed: bool, map_: Map, w: float) -> None:
		VCFHMM.__init__(self, records, map_)
		self.records: list[VCFFamilyRecord] = records
		self.ref_haps = ref_haps
		self.is_mat_imputed = is_mat_imputed
		self.prev_h_table = self.collect_possible_previous_hidden_states()
		# crossover values
		# 後代
		self.Cc = [ Map.Kosambi(self.dist(r1, r2))
							for r1, r2 in zip(records, records[1:]) ]
		K = 5						# 親の遷移確率の子に対する倍率
		# 親は直接の後代でないので、乗り換え確率を高くする
		self.Cp = [ Map.Kosambi(self.dist(r1, r2) * K)
							for r1, r2 in zip(self.records, self.records[1:]) ]
	
	def NH(self) -> int:
		return len(self.ref_haps)
	
	def num_progenies(self) -> int:
		return len(self.records[0].v) - 11
	
	# ハプロタイプを決めたときのGenotype
	def gt_by_haplotypes(self, hc1: int, hc2: int,
								mat_gt: int, pat_gt: int) -> int:
		return ((mat_gt >> hc1) & 1) | (((pat_gt >> hc2) & 1) << 1)
	
	# i: record index, j: progeny index
	def emission_probability(self, i: int, h: int, op: int, ocs: list[int],
												phased_parent_gt: int) -> float:
		N = self.num_progenies()
		hp = h >> (N*2)
		hp2, hp1 = divmod(hp, self.NH())
		non_phased_parent_gt = (self.ref_haps[hp1][i] |
								(self.ref_haps[hp2][i] << 1))
		mat_gt = (phased_parent_gt if self.is_mat_imputed
										else non_phased_parent_gt)
		pat_gt = (non_phased_parent_gt if self.is_mat_imputed
										else phased_parent_gt)
		Ep = self.E[non_phased_parent_gt][op]	# parent emission
		# 後代の排出確率を計算
		Ec = 0.0
		for j in range(N):
			hc2, hc1 = divmod((h >> (j * 2)) & 3, 2)
			gtc = self.gt_by_haplotypes(hc1, hc2, mat_gt, pat_gt)
			Ec += self.E[gtc][ocs[j]]
		return Ep + Ec
	
	def progeny_transition_probability(self, i: int,
											prev_hc: int, hc: int) -> float:
		cc = self.Cc[i-1]	# 後代の遷移確率
		hc1 = hc & 1
		hc2 = hc >> 1
		prev_hc1 = prev_hc & 1
		prev_hc2 = prev_hc >> 1
		# 遷移確率 0なら(1-c)、1ならcを掛ける
		return (log(cc if prev_hc1 != hc1 else 1.0 - cc) +
				log(cc if prev_hc2 != hc2 else 1.0 - cc))
	
	def initialize_dp(self) -> list[DP]:
		M = len(self.records)	# マーカー数
		L = self.NH()**2 << (2*self.num_progenies())
		dp = [ [ (MIN_PROB, 0) ] * L for _ in range(M) ]
		phased_col = 9 if self.is_mat_imputed else 10
		non_phased_col = 10 if self.is_mat_imputed else 9
		record = self.records[0]
		phased_parent_gt = Genotype.phased_gt_to_int(record.v[phased_col])
		op = Genotype.gt_to_int(record.v[non_phased_col])	# observed parent
		# observed progs
		ocs = [ Genotype.gt_to_int(gt) for gt in record.v[11:] ]
		for h in range(L):		# hidden state
			E_all = self.emission_probability(0, h, op, ocs, phased_parent_gt)
			dp[0][h] = (E_all, h)
		return dp
	
	# hidden stateに対して、可能な前のhidden stateを集めておく
	def collect_possible_previous_hidden_states(self) -> list[list[int]]:
		N = self.num_progenies()
		L = self.NH()**2 << (2*N)
		Lc = 1 << (2*N)				# 子のハプロタイプの状態数
		prev_h_table: list[list[int]] = [ [] for _ in range(L) ]
		for h in range(L):		# hidden state
			hc = h & (Lc - 1)
			hp = h >> (N*2)
			hp2, hp1 = divmod(hp, self.NH())
			# 両側乗り越えることはないとする
			# non-phasedの親のあり得る前の状態
			prev_hps = []
			for h1 in range(self.NH()):
				prev_h1 = h1 + hp2 * self.NH()
				prev_hps.append(prev_h1)
			for h2 in range(self.NH()):
				if h2 == hp2:	# for duplication
					continue
				prev_h1 = hp1 + h2 * self.NH()
				prev_hps.append(prev_h1)
			
			# 後代のあり得る前の状態
			prev_hcs = []
			for hc1 in range(1 << (N*2)):
				t = hc ^ hc1		# 乗り換えたかをビットで表す
				# 乗り換えは合わせて1回まで許容できる
				if sum(1 for k in range(N*2) if ((t >> k) & 1) == 1) <= 1:
					prev_hcs.append(hc1)
			
			for prev_hp, prev_hc in product(prev_hps, prev_hcs):
				if prev_hp == hp or prev_hc == hc:
					prev_h = (prev_hp << (N*2)) | prev_hc
					prev_h_table[h].append(prev_h)
		
		return prev_h_table
	
	def compute_non_phased_parent_gt(self, h: int, i: int) -> int:
		N = self.num_progenies()
		hp = h >> (N*2)
		hp2, hp1 = divmod(hp, self.NH())
		return self.ref_haps[hp1][i] | (self.ref_haps[hp2][i] << 1)
	
	def update_dp(self, i: int, dp: list[DP]) -> None:
		N = self.num_progenies()
		L = self.NH()**2 << (2*N)
		Lc = 1 << (2*N)				# 子のハプロタイプの状態数
		record = self.records[i]
		phased_col = 9 if self.is_mat_imputed else 10
		non_phased_col = 10 if self.is_mat_imputed else 9
		cc = self.Cc[i-1]	# 遷移確率
		cp = self.Cp[i-1]	# 親の遷移確率
		phased_parent_gt = Genotype.phased_gt_to_int(record.v[phased_col])
		op = Genotype.gt_to_int(record.v[non_phased_col])	# observed parent
		# observed progs
		ocs = [ Genotype.gt_to_int(gt) for gt in record.v[11:] ]
		
		for h in range(L):		# hidden state
			E_all = self.emission_probability(i, h, op, ocs, phased_parent_gt)
			
			hc = h & (Lc - 1)
			hp2, hp1 = divmod(h >> (N*2), self.NH())
			
			for prev_h in self.prev_h_table[h]:
				t1 = prev_h ^ hc	# 乗り換えしていないかしたかをbitで
				# 遷移確率 0なら(1-c)、1ならcを掛ける
				Tc = 0.0
				for j in range(N*2):
					Tc += log(cc if ((t1 >> j) & 1) == 1 else 1.0 - cc)
				# 親側の遷移
				prev_hp2, prev_hp1 = divmod(prev_h >> (N * 2), self.NH())
				Tp = (log(cp if prev_hp1 != hp1 else 1.0 - cp) +
					  log(cp if prev_hp2 != hp2 else 1.0 - cp))
				
				prob = dp[i-1][prev_h][0] + (Tc + Tp + E_all)
				dp[i][h] = max(dp[i][h], (prob, prev_h))
	
	def update_genotypes(self, hs: list[int]) -> None:
		M = len(self.records)
		N = self.num_progenies()
		for i in range(M):
			record = self.records[i]
			if self.is_mat_imputed:
				pat_gt = self.compute_non_phased_parent_gt(hs[i], i)
				record.set_GT(1, Genotype.int_to_phased_gt(pat_gt))
				mat_gt = Genotype.phased_gt_to_int(record.v[9])
			else:
				mat_gt = self.compute_non_phased_parent_gt(hs[i], i)
				record.set_GT(0, Genotype.int_to_phased_gt(mat_gt))
				pat_gt = Genotype.phased_gt_to_int(record.v[10])
			for j in range(N):	# 個々の後代
				hc2, hc1 = divmod((hs[i] >> (j * 2)) & 3, 2)
				gtc_int = self.gt_by_haplotypes(hc1, hc2, mat_gt, pat_gt)
				record.set_GT(j+2, Genotype.int_to_phased_gt(gtc_int))
	
	def impute(self) -> None:
		# DP
		# 外側から、マーカー、ハプロタイプの状態
		M = len(self.records)
		dp = self.initialize_dp()
		for i in range(1, M):
			self.update_dp(i, dp)
		
		hs = self.trace_back(dp)
		self.update_genotypes(hs)
